app/helper.py

```diff
@@ -0,0 +1,66 @@
+"""Url shortener helper to help implement various functionalities.
+
+Helps keep the code base DRY.
+"""
+import random
+import string
+
+from flask import abort
+from .models import ShortUrl, LongUrl
+from app import db
+
+
+class UrlSaver(object):
+    """Handles the various functionalities implemented."""
+
+    @staticmethod
+    def generate_and_save_urls(url, user, vanity_string=None):
+        """Help manage the generating and saving of urls."""
+        if url in [x.long_url for x in user.long_urls]:
+            long_url = LongUrl.query.filter_by(long_url=url).first()
+            return [short_url for short_url in
+                    user.short_urls if short_url.long_url_id == long_url.id][0]
+
+        elif vanity_string:
+            short_url = ShortUrl.query.filter_by(
+                                    short_url=vanity_string).first()
+            if short_url:
+                abort(400, "Vanity string already in use. Pick another.")
+            save_data = UrlSaver.save_url(vanity_string, url, user)
+            return save_data
+        else:
+            short_url = UrlSaver.generate_short_url()
+            save_data = UrlSaver.save_url(short_url, url, user)
+            return save_data
+
+    @staticmethod
+    def save_url(short_url, long_url, user):
+        """Save and append short and long urls to their various tables."""
+        long_url_exist = LongUrl.query.filter_by(long_url=long_url).first()
+
+        if long_url_exist:
+            shorturl = ShortUrl(short_url=short_url,
+                                long_url_id=long_url_exist.id)
+            user.short_urls.append(shorturl)
+            long_url_exist.users.append(user)
+            db.session.commit()
+            return shorturl
```

Comment:
We can find common logic with both cases that can be moved right after the condition.